### PR TITLE
Refactor Tactics to Take Dynamic Parameters Config

### PR DESCRIPTION
**Pull request by jonathanlew**
### Description
* refactor tactics to take dynamic parameter config
* refactor goalie tactic

### Testing Done
regression tests pass

### Resolved Issues
#1860

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/ai/hl/stp/tactic/goalie_tactic.cpp [hunk 1]
```diff
@@ -152,155 +153,161 @@ void GoalieTactic::calculateNextAction(ActionCoroutine::push_type &yield)
...
+{
+    // if the ball is in the don't chip rectangle we do not chip the ball
+    // as we risk bumping the ball into our own net trying to move behind
+    // the ball
+    auto dont_chip_rectangle =
+        Rectangle(field.friendlyGoalpostNeg(),
+                  field.friendlyGoalpostPos() + Vector(2 * ROBOT_MAX_RADIUS_METERS, 0));
+
+    // if the ball is slow but its not safe to chip it out, don't.
+    // TODO finesse the ball out of the goal using the dribbler.
```
Comment (kelvinkoon) [inline src/software/ai/hl/stp/tactic/goalie_tactic.cpp]: TODO should link to an issue.

File: src/software/ai/hl/stp/tactic/goalie_tactic.cpp [hunk 2]
```diff
@@ -152,155 +153,161 @@ void GoalieTactic::calculateNextAction(ActionCoroutine::push_type &yield)
...
+    move_action->updateControlParams(*robot, goalie_pos, goalie_orientation, 0.0,
+                                     DribblerEnable::OFF, MoveType::NORMAL,
+                                     AutochickType::AUTOCHIP, BallCollisionType::ALLOW);
+    return move_action;
+}
+
+std::shared_ptr<Action> GoalieTactic::chipBallIfSafe(
+    std::shared_ptr<ChipAction> chip_action, std::shared_ptr<StopAction> stop_action)
+{
+    // if the ball is in the don't chip rectangle we do not chip the ball
```
Comment (kelvinkoon) [inline src/software/ai/hl/stp/tactic/goalie_tactic.cpp]: "if the ball is in the __" seems to be missing a word.

**Review by kelvinkoon — CHANGES_REQUESTED** [review]
LGTM overall!

File: src/software/ai/hl/stp/tactic/shoot_goal_tactic.cpp [hunk 3]
```diff
@@ -74,14 +75,11 @@ bool ShootGoalTactic::isEnemyAboutToStealBall() const
...
-                                     ->EnemyAboutToStealBallRectangleWidth()
-                                     ->value();
-    auto steal_ball_rect_length = DynamicParameters->getAIConfig()
-                                      ->getShootGoalTacticConfig()
-                                      ->EnemyAboutToStealBallRectangleExtensionLength()
-                                      ->value();
+    auto steal_ball_rect_width =
+        shoot_goal_tactic_config->EnemyAboutToStealBallRectangleWidth()->value();
+    auto steal_ball_rect_length =
+        shoot_goal_tactic_config->EnemyAboutToStealBallRectangleExtensionLength()
```
Comment (Resocram) [inline src/software/ai/hl/stp/tactic/shoot_goal_tactic.cpp:81]: What is the extension length for?

File: src/software/ai/hl/stp/tactic/goalie_tactic.cpp (hunk 1 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/tactic/goalie_tactic.cpp]: clarified

File: src/software/ai/hl/stp/tactic/goalie_tactic.cpp (hunk 2 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/tactic/goalie_tactic.cpp]: clarified

File: src/software/ai/hl/stp/tactic/shoot_goal_tactic.cpp (hunk 3 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/tactic/shoot_goal_tactic.cpp:81]: it's used to define the length of the rectangle in front of the robot where if an enemy robot is inside, we consider it a threat

**Review by kelvinkoon — APPROVED** [review]
LGTM 💯

**Review by Resocram — APPROVED** [review]

**Review by akhilveeraghanta — APPROVED** [review]

**Pull request merged**